### PR TITLE
CodeMirror 5.36.0 modularized

### PR DESCRIPTION
**Pull request by BurningTreeC**
#3169

### CODE REVIEW

**Comment by BurningTreeC** [comment]
@Jermolene 

List of improvements:

- Upgrade version to 5.36.0
- Modularization into
    - Base plugin
         - basic lightweigth configuration, ~230kb
    - "AddOn" plugins
        - Search and Replace
        - Autocompletion
        - Fullscreen editing
        - Auto-Close brackets
        - Auto-Close tags
        - Sublime Text keymap
        - Emacs keymap
        - Vim keymap
        - Highlighting modes
            - Javascript
            - Xml
            - Html (htmlmixed)
           - Html (htmlembedded)
           - Css
           - Markdown
            - X-Tiddlywiki
- Automatic configuration
- ControlPanel configuration tab
    - line numbers
    - theme
     - editor fontfamily (monospace or sans-serif)
     - cursor settings
     - line wrapping
    - active line highlighting
     - autocompletion on/off (if installed)

@pmario - anything missing?

**Comment by BurningTreeC** [comment]
@pmario 

do we still want auto-installation like when installing an addon and the base plugin is missing or when installing htmlmixed mode which depends on xml, css and javascript?

**Comment by BurningTreeC** [comment]
@pmario 

we should also change the default and extra keymaps for

- "Esc"
- "Ctrl+Enter"
- Toolbar shortcuts
- "Ctrl+T" and others ... shortcuts that are handled by the browser

so that they don't interfere

File: plugins/tiddlywiki/codemirror-keymap-sublime-text/config/extraKeysTW.tid [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  "Shift-Enter": "insertLineAfter",
+  "Esc": "",
+  "Ctrl-Esc": "singleSelectionTop",
+  "Ctrl-L": "",
+  "Shift-Ctrl-L": "",
+  "Alt-L": "selectLine",
+  "Shift-Alt-L": "splitSelectionByLine",
+  "Ctrl-T": "",
+	"Alt-T": "transposeChars",
+	"Ctrl-K Ctrl-U": "",
```
Comment (pmario) [inline plugins/tiddlywiki/codemirror-keymap-sublime-text/config/extraKeysTW.tid]: indentation ? space and tab mixed ??

Comment (BurningTreeC) [reply plugins/tiddlywiki/codemirror-keymap-sublime-text/config/extraKeysTW.tid]: that's possible, copy-pasta

**Comment by BurningTreeC** [comment]
@pmario 

I will adjust the vim and emacs keymaps a bit later
Not all my shortcut-choices may be the best in terms of comfort/accessibility but there isn't that much scope to play with
I hope I got all working correctly for the basic and sublime keymap

Do you have an idea about dependency-installation? should I try what I suggested in the previous thread?

**Comment by pmario** [comment]
TW core has a dependency mechanism, in the `plugin.info` ... It's the `dependents` field. ... strange name but anyway. .. The problem is, imo it isn't used at the moment. It's just there.... 

So we would need to implement something. ..

**Comment by BurningTreeC** [comment]
@pmario 

>To be honest. ... I don't think, this PR will ever be merged, if you want to have it 100% right. ... eg: vnd/tiddlywiki will need a mixed mode ... and so on.

I'm fine with leaving it as it is, making the last changes on the vim and emacs keymaps and then giving it a go

**Comment by BurningTreeC** [comment]
Ok, the vim and emacs keymaps are  too cryptic to make changes now

I want to leave them how they are, that can also be another PR at one point

@pmario - let's merge it

**Comment by BurningTreeC** [comment]
@Jermolene 

This is ready to merge.

**Comment by BurningTreeC** [comment]
@pmario ,

I've just added a fix in engine.js for the focus-loosing when clicking the toolbar buttons 
There are some Todos and comments left from your auto-config pr ...
Can you tell me what's obsolete and can be removed?

Thanks, Simon

**Comment by pmario** [comment]
>There are some Todos and comments left from your auto-config pr ...

There are 3. They can be removed. ... The only thing, that could be done, test some "broken" config values and see if the program goes wild. ...

**Comment by pmario** [comment]
The first indent level is 0  ... So the outer wrapper doesn't count!

**Comment by BurningTreeC** [comment]
@pmario @Jermolene 

can this be considered for the next release?

@pmario 

I've removed the comments and I think I got the formatting right
Why are you telling me I should test broken config values when it's you who implemented it that way?
It works fine overall but there's no dependency check, yeah... things will be broken if things are missing

If you want me to tell you which addon has other addons as dependency, so that YOU can add a dependency check in engine.js and don't add it to the config, only send a notification message like it's done in another plugin - then that's fine - just communicate.

Simon

**Comment by pmario** [comment]
>Why are you telling me I should test broken config values when it's you who implemented it that way?

We'll see, how users are going to break the config stuff. Then we'll fix it.

**Comment by Jermolene** [comment]
> can this be considered for the next release?

Yes indeed. I'd love to get this merged. It's been a monster collaboration between you both to get this done, and many thanks for your patience and attention to detail. Congratulations!

**Comment by BurningTreeC** [comment]
@Jermolene 

Thank you very much and thank you @pmario !

> We'll see, how users are going to break the config stuff. Then we'll fix it.

That's good, Thanks again!

**Comment by pmario** [comment]
>... and many thanks for your patience and attention to detail. Congratulations!

That's right. Well done. .. I think we have a relatively easy to use base system now. ... I like "search / replace" quite a lot. ...

**Comment by BurningTreeC** [comment]
@Jermolene 

List of improvements:

- Upgrade version to 5.36.0
- Modularization into
    - Base plugin
         - basic lightweigth configuration, ~240kb
    - "AddOn" plugins
        - Search and Replace
        - Autocompletion
        - Fullscreen editing
        - Auto-Close brackets
        - Auto-Close tags
        - Sublime Text keymap
        - Emacs keymap
        - Vim keymap
        - Highlighting modes
            - Javascript
            - Xml
            - Html (htmlmixed)
           - Html (htmlembedded)
           - Css
           - Markdown
            - X-Tiddlywiki
- Automatic configuration
    - the configuration tiddler isn't needed anymore
        - codemirror (engine.js) builds the configuration using information within the installed cm plugins
- ControlPanel configuration tab
    - line numbers
    - theme
     - editor fontfamily (same as tw-native)
     - cursor settings
     - line wrapping
    - active line highlighting
     - autocompletion on/off (if installed)
    - keymap

**Comment by Jermolene** [comment]
Great, thanks @BurningTreeC

**Comment by Jermolene** [comment]
Hi @BurningTreeC it looks like CodeMirror v5.37.0 is now out: https://github.com/codemirror/CodeMirror/releases/tag/5.37.0

**Pull request merged**